reflect.go

```diff
@@ -0,0 +1,322 @@
+func (rm *reflectedMockCall) CallReturnEmpty(retErr error) {
+	var (
+		returnType = rm.returnMethod.Type()
+		returnLen  = returnType.NumIn()
+		emptyArgs  = make([]reflect.Value, 0)
+	)
+
+	if returnLen == 1 && returnType.In(0).Name() == "" && retErr != nil {
+		// Special case: we have an error to return and one argument with an unknown
+		// type. This can happen in two cases. First, when we're handling a bare
+		// testify mock and don't know the Return type. Secondly, when we're handling
+		// a mockery mock with a custom Return type which will be hidden to us. In
+		// the first case, we want to call Return(retErr). The second case is a user
+		// error and will panic.
+		emptyArgs = append(emptyArgs, reflect.ValueOf(retErr))
+	} else {
+		for i := 0; i < returnLen; i++ {
+			argType := returnType.In(i)
+
+			if retErr != nil && argType.Name() == "error" {
+				// We were given an error to return -- use it!
+				emptyArgs = append(emptyArgs, reflect.ValueOf(retErr))
+			} else if !returnType.IsVariadic() || i < returnLen-1 {
+				emptyArgs = append(emptyArgs, reflect.Zero(argType))
+			}
+		}
+	}
+
+	rm.returnMethod.Call(emptyArgs)
+}
+
+// CallReturn calls the mock's Return method with the given args.
+//
+// If an optional retErr is provided, we will use that for error values.
+func (rm *reflectedMockCall) CallReturn(args []any, retErr error, zeroValueErrs bool) error {
+	var (
+		returnType = rm.returnMethod.Type()
+		returnLen  = returnType.NumIn()
+		returnArgs = append([]any{}, args...)
+	)
+
+	if retErr != nil {
+		// We were given an error to return -- use it!
+		returnArgs = append(returnArgs, retErr)
+	} else if zeroValueErrs {
+		// Add a nil error, if applicable
+		for i := len(args); i < returnLen; i++ {
+			if returnType.In(i).Name() == "error" {
+				var emptyErr error
+				returnArgs = append(returnArgs, emptyErr)
+				break
+			}
+		}
+	}
+
+	rm.mustArgMatch(returnType, returnArgs)
+
+	rargs, err := toReflectValues(returnArgs, returnType)
+	if err != nil {
+		return fmt.Errorf("toReflectValues failed to transform return values: %s", err)
+	}
+
+	rm.returnMethod.Call(rargs)
+	return nil
+}
+
+// toReflectValues transforms the |args| of the |method| from `[]any` to
+// `[]reflect.Value`.
+func toReflectValues(args []any, typ reflect.Type) ([]reflect.Value, error) {
+	if len(args) != typ.NumIn() && !typ.IsVariadic() {
+		return nil, fmt.Errorf(
+			"mismatched number of args: expected %d but got %d",
+			typ.NumIn(),
+			len(args),
+		)
+	}
+
+	values := make([]reflect.Value, len(args))
+
+	for i, arg := range args {
+		argType := typ.In(min(i, typ.NumIn()-1))
+
+		if arg != nil {
+			values[i] = reflect.ValueOf(arg)
+		} else {
+			switch argType.Kind() {
+			case reflect.Interface, reflect.Ptr:
+				values[i] = reflect.Zero(argType)
+			default:
+				return nil, fmt.Errorf(
+					"cannot handle nil for non-interface or non-pointer type: %s",
+					argType,
+				)
+			}
```

Comment:
```suggestion
			switch argType.Kind() {
			case reflect.Interface, reflect.Ptr, reflect.Slice:
				values[i] = reflect.Zero(argType)
			default:
				return nil, fmt.Errorf(
					"cannot handle nil for non-interface, non-pointer, non-slice type: %s",
					argType,
				)
			}
```
Something like this would fix the problem I described in my comment above (I reran the test and it passed) but I have no idea if this would be logically incorrect for some other reason.

Reply:
oooooooo